distributed/deploy/spec.py

```diff
@@ -0,0 +1,296 @@
+        self,
+        workers=None,
+        scheduler=None,
+        worker=None,
+        asynchronous=False,
+        loop=None,
+        silence_logs=False,
+    ):
+        self._created = weakref.WeakSet()
+        if scheduler is None:
+            try:
+                from distributed.bokeh.scheduler import BokehScheduler
+            except ImportError:
+                services = {}
+            else:
+                services = {("bokeh", 8787): BokehScheduler}
+            scheduler = {"cls": Scheduler, "options": {"services": services}}
+
+        self.scheduler_spec = scheduler
+        self.worker_spec = workers or {}
+        self.new_spec = worker
+        self.workers = {}
+        self._i = 0
+        self._asynchronous = asynchronous
+
+        if silence_logs:
+            self._old_logging_level = silence_logging(level=silence_logs)
+
+        self._loop_runner = LoopRunner(loop=loop, asynchronous=asynchronous)
+        self.loop = self._loop_runner.loop
+
+        self.scheduler = self.scheduler_spec["cls"](
+            loop=self.loop, **self.scheduler_spec["options"]
+        )
+        self.status = "created"
+
+        if not self.asynchronous:
+            self._loop_runner.start()
+            self.sync(self._start)
+
+        self._correct_state_waiting = None
+
+    async def _start(self):
+        while self.status == "starting":
+            await asyncio.sleep(0.01)
+        if self.status == "running":
+            return
+        if self.status == "closed":
+            raise ValueError("Cluster is closed")
+
+        self._lock = asyncio.Lock()
+        self.status = "starting"
+        self.scheduler = await self.scheduler
+        self.status = "running"
+
+    def _correct_state(self):
+        if self._correct_state_waiting:
+            # If people call this frequently, we only want to run it once
+            return self._correct_state_waiting
+        else:
+            task = asyncio.Task(self._correct_state_internal())
+            self._correct_state_waiting = task
+            return task
+
+    async def _correct_state_internal(self):
+        async with self._lock:
+            self._correct_state_waiting = None
+
+            pre = list(set(self.workers))
+            to_close = set(self.workers) - set(self.worker_spec)
+            if to_close:
+                await self.scheduler.retire_workers(workers=list(to_close))
+                tasks = [self.workers[w].close() for w in to_close]
+                for task in tasks:
+                    await task
+            for name in to_close:
+                del self.workers[name]
+
+            to_open = set(self.worker_spec) - set(self.workers)
+            workers = []
+            for name in to_open:
+                d = self.worker_spec[name]
+                cls, opts = d["cls"], d.get("options", {})
+                if "name" not in opts:
+                    opts = toolz.merge({"name": name}, opts, {"loop": self.loop})
```

Comment:
Did you mean to include the `loop` in here?

Reply:
Yes, ideally we want the worker to use the IOLoop used by the cluster object.

Reply:
I mean that `loop` is only added if `name` is not in opts. Wouldn't you always want to pass it?

Reply:
Ah, indeed.  Looking at this again it looks like we do this in an async def function anyway, so `IOLoop.current()` should be valid regardless.  I'll remove the reference to loop entirely, which should be helpful in reducing the contract too.